### PR TITLE
fix(sqllab): excessive API calls for schemas

### PR DESCRIPTION
**Pull request by justinpark**
### SUMMARY
Currently, the refresh logic for schemas is being triggered excessively due to unnecessary dependencies, causing it to be called every time a tab is switched.

This commit addresses the issue by modifying the logic to execute only when a trigger is needed.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Before:

https://github.com/apache/superset/assets/1392866/b0ba7b17-1c7a-4d0e-aa9b-1e2fcaa5604b

After:

https://github.com/apache/superset/assets/1392866/a785f3fe-3d04-4c1b-a574-34783d6c496c

### TESTING INSTRUCTIONS

Switch the tabs between the same db selected tabs
Check the network tab and then only schemas called once

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/hooks/apiResources/catalogs.ts [hunk 1]
```diff
@@ -88,6 +87,10 @@ export function useCatalogs(options: Params) {
     onError?.();
   });
 
+  const resolver = useEffectEvent(() =>
```
Comment (michael-s-molina) [inline superset-frontend/src/hooks/apiResources/catalogs.ts:90]: There's a lot of duplicated code. Maybe something like this?

```
export function useCatalogs(options: Params) {
  const { dbId, onSuccess, onError } = options || {};
  const [trigger] = useLazyCatalogsQuery();
  const result = useCatalogsQuery(
    { dbId, forceRefresh: false },
    {
      skip: !dbId,
    },
  );

  const fetchData = useCallback(
    (forceRefresh: boolean) => {
      trigger({ dbId, forceRefresh }).then(({ isSuccess, isError, data }) => {
        if (isSuccess) {
          onSuccess?.(data || EMPTY_CATALOGS, forceRefresh);
        }
        if (isError) {
          onError?.();
        }
      });
    },
    [dbId, onError, onSuccess, trigger],
  );

  const refetch = useCallback(() => {
    if (dbId) {
      fetchData(true);
    }
  }, [dbId, fetchData]);

  useEffect(() => {
    if (dbId && !result.currentData) {
      fetchData(false);
    }
  }, [dbId, fetchData, result.currentData]);

  return {
    ...result,
    refetch,
  };
}
```

File: superset-frontend/src/hooks/apiResources/schemas.ts [hunk 2]
```diff
@@ -92,20 +91,24 @@ export function useSchemas(options: Params) {
     onError?.();
   });
 
+  const resolver = useEffectEvent(() =>
```
Comment (michael-s-molina) [inline superset-frontend/src/hooks/apiResources/schemas.ts:94]: Same as before.

**Review by michael-s-molina — COMMENTED** [review]
Thanks for the PR @justinpark!

File: superset-frontend/src/hooks/apiResources/catalogs.ts (hunk 1 above)
Comment (justinpark) [reply superset-frontend/src/hooks/apiResources/catalogs.ts:90]: That's a good suggestion. One thing that including result.currentData in the dependencies array of useEffect can cause unintended fetchData calls. I'll change it to handle this within the fetchData function.

**Review by michael-s-molina — APPROVED** [review]

**Pull request merged**